### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Ivan-Nieto**
Finished bank app with the following routes.

- GET /accounts
- GET /accounts/:id
- GET /accounts/:id/transactions
- GET /transactions

- POST /accounts
- POST /accounts/:id/transactions/add
- POST /accounts/:id/transactions/withdraw
- POST /accounts/:id/transactions/send

Data validation for each route is handled by custom pipes and business logic is handled by the service.

### CODE REVIEW

File: src/accounts/accounts.controller.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+@Controller('accounts')
+export class AccountsController {
+  constructor(private accountService: AccountsService) {}
+
+  @Get()
+  public getAccounts() {
+    return this.accountService.getAccounts();
+  }
+
+  @Get(':id')
```
Comment (cyrusBaker) [inline src/accounts/accounts.controller.ts:22]: If you have an id the path should indicate what type of id it is. Eg. accounts/:id

**Review by cyrusBaker — APPROVED** [review]

File: src/accounts/accountValidationPipe.ts [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+      (e) => e.id === value.target_account_id,
+    );
+    if (!targetAccount)
+      throw new BadRequestException('Target account not found');
+
+    return value;
+  }
+}
+
+export default AccountValidationPipe;
```
Comment (cyrusBaker) [inline src/accounts/accountValidationPipe.ts:175]: Could you add a little more validation? Checks for a negative account balance upon creation, check for valid currency string, check for valid email. Some of these can be done in the dto using the class-validator library.

File: src/accounts/accounts.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    email_address: 'Jane.Doe@example.com',
+    id: '0004',
+    balance: {
+      amount: 47,
+      currency: 'USD',
+    },
+    note: 'a customer',
+  },
+];
+export default ACCOUNTS;
```
Comment (cyrusBaker) [inline src/accounts/accounts.ts:47]: Nice touch seeding the accounts

File: src/accounts/Account.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    if (invalidString(id)) return false;
+    return this.accounts.every((e) => e.id !== id);
+  }
+
+  public exists(id: string) {
+    if (invalidString(id)) return false;
+    return this.accounts.some((e) => e.id === id);
+  }
+
+  public toObject(): AccountDto {
```
Comment (MRdgz) [inline src/accounts/Account.ts:144]: It's good to see that you created a function for mapping. Our current practice for this is to create a mapper class per domain model within this class we handle converting DTOs to Entities and vice-versa . [Example](https://github.com/electronic-caregiver/svcs-api-vitals/blob/dev/src/reminders/mappers/reminder-messages.map.ts)

File: src/accounts/accounts.service.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+          targetAccount.balance?.currency || config.amount_money.currency,
+        amount:
+          (targetAccount.balance?.amount || 0) + config.amount_money.amount,
+      },
+    };
+
+    // Add transaction record
+    const { error } = trans.save();
+    if (error) {
+      // Revert both accounts
```
Comment (MRdgz) [inline src/accounts/accounts.service.ts:152]: Nice!

**Review by MRdgz — APPROVED** [review]

**Pull request merged**